### PR TITLE
Ogg reader

### PR DESCRIPTION
**Pull request by LVala**

### CODE REVIEW

File: examples/send_from_file/README.md [hunk 1]
```diff
@@ -1,7 +1,10 @@
...
+Send video and audio from files to a browser.
 
 1. Start `ex_ice/signalling_server` with `mix run --no-halt`
 2. Run `elixir example.exs`
-3. Visit `example.html` in your browser e.g. `file:///home/Repos/elixir-webrtc/ex_webrtc/examples/send_from_file/example.html`
\ No newline at end of file
+3. Visit `example.html` in your browser e.g. `file:///home/Repos/elixir-webrtc/ex_webrtc/examples/send_from_file/example.html`
+4. Press the play button.
+
+You can replace `video.avf` or `audio.ogg` and use your own files instead.
```
Comment (mickel8) [inline examples/send_from_file/README.md:10]: video.avf -> video.ivf

File: examples/send_from_file/example.exs [hunk 2]
```diff
@@ -111,32 +123,64 @@ defmodule Peer do
...
+
+  @impl true
+  def handle_info(:send_audio_packet, state) do
+    case OggReader.next_packet(state.audio_reader) do
+      {:ok, reader, {packet, duration}} ->
+        # in real-life scenario, you will need to conpensate for `Process.send_after/3` error
+        # and time spent on reading and parsing the file
+        # that's why you might hear short pauses in audio playback, when using this example
+        Process.send_after(self(), :send_audio_packet, duration)
+        rtp_packet = ExRTP.Packet.new(packet, 111, 1000, state.last_audio_timestamp, 1000)
```
Comment (mickel8) [inline examples/send_from_file/example.exs:162]: I would pass 0 for everything that RTP sender is responsible for. Right now, we might confuse user that they are responsible for handling e.g. pt

File: examples/send_from_file/example.exs [hunk 3]
```diff
@@ -111,32 +123,64 @@ defmodule Peer do
...
+        {:ok, reader} = IVFReader.open("./video.ivf")
+        {:ok, _header} = IVFReader.read_header(reader)
+        {:noreply, %{state | video_reader: reader}}
+    end
+  end
+
+  @impl true
+  def handle_info(:send_audio_packet, state) do
+    case OggReader.next_packet(state.audio_reader) do
+      {:ok, reader, {packet, duration}} ->
```
Comment (mickel8) [inline examples/send_from_file/example.exs:157]: The convention is to return reader at the last place in the result tuple. Take a look at Map.pop and others

File: examples/send_from_file/example.exs [hunk 4]
```diff
@@ -151,13 +195,18 @@ defmodule Peer do
          %{"role" => _role, "type" => "peer_joined"},
          %{peer_connection: pc} = state
        ) do
-    track = MediaStreamTrack.new(:video)
-    {:ok, _} = PeerConnection.add_transceiver(pc, track, codec: :vp8)
+    audio_track = MediaStreamTrack.new(:audio)
+    video_track = MediaStreamTrack.new(:video)
+
+    {:ok, _} = PeerConnection.add_track(pc, audio_track)
```
Comment (mickel8) [inline examples/send_from_file/example.exs:201]: Brilliant!

File: examples/send_from_file/example.exs [hunk 5]
```diff
@@ -202,13 +251,16 @@ defmodule Peer do
...
     {:ok, _header} = IVFReader.read_header(ivf_reader)
-    payloader = VP8Payloader.new(800)
+    vp8_payloader = VP8Payloader.new(800)
+
+    {:ok, ogg_reader} = OggReader.open("./audio.ogg")
 
-    Process.send_after(self(), :send_frame, 30)
-    %{state | ivf_reader: ivf_reader, payloader: payloader}
+    Process.send_after(self(), :send_video_frame, 30)
+    Process.send_after(self(), :send_audio_packet, 20)
```
Comment (mickel8) [inline examples/send_from_file/example.exs:262]: Actually, we could do Process.sendafter 0

File: lib/ex_webrtc/media/ogg_reader.ex [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+  For now, works only with single Opus stream in the container.
+  This function reads the ID and Comment Headers (and, for now, ignores them).
+  """
+  @spec open(Path.t()) :: {:ok, t()} | {:error, File.posix() | :invalid_header}
+  def open(path) do
+    with {:ok, file} <- File.open(path),
+         reader <- %{file: file, packets: [], rest: <<>>},
+         # for now, we ignore ID Header and Comment Header
+         {:ok, reader, <<@id_signature, _rest::binary>>} <- do_next_packet(reader),
```
Comment (mickel8) [inline lib/ex_webrtc/media/ogg_reader.ex:46]: As above, keep reader at the end of return tuple

File: lib/ex_webrtc/media/ogg_reader.ex [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+      {:error, _res} = err -> err
+      _other -> {:error, :invalid_header}
+    end
+  end
+
+  @doc """
+  Reads next Ogg packet.
+
+  One Ogg packet is equivalent to one Opus packet.
+  This function also returns the duration of the audio in milliseconds, based on Opus packet TOC sequence.
```
Comment (mickel8) [inline lib/ex_webrtc/media/ogg_reader.ex:59]: Maybe let's elaborate on TOC

**Review by mickel8 — APPROVED** [review]
I didn't dive into Ogg itself but other than that looks great! I just added a couple of minor suggestions

File: examples/send_from_file/example.exs (hunk 5 above)
Comment (LVala) [reply examples/send_from_file/example.exs:262]: Even better, `send(self(), ...)` 🙂

File: lib/ex_webrtc/media/ogg_reader.ex (hunk 7 above)
Comment (LVala) [reply lib/ex_webrtc/media/ogg_reader.ex:59]: I'll just redirect the reader to the RFC.

**Pull request merged**